lib/Normalize_std_ast.ml

```diff
@@ -216,61 +212,116 @@ let make_mapper conf ~ignore_doc_comments ~erase_jane_syntax =
-          ; _ }
-        , expression )
-      when erase_jane_syntax ->
-        let default_pos = dummy_position ~loc in
-        let expression =
-          let pexp_desc =
-            Pexp_fun (Optional l, Some default_pos, pat, expression)
+          { exp with
+            pexp_attributes= attrs
+          ; pexp_desc=
+              Pexp_function
+                ( [ { pparam_loc= l.loc
+                    ; pparam_desc= Pparam_newtype (l, Some jkind) } ]
+                , Some {mode_annotations= []; type_constraint= Pconstraint ty}
+                , Pfunction_body exp1 ) }
+    | _ -> (
+      match pexp_desc with
+      | Pexp_apply
+          ( {pexp_desc= Pexp_extension ({txt= "extension.exclave"; _}, _); _}
+          , [(Nolabel, expr)] )
+        when erase_jane_syntax ->
+          m.expr m expr
+      | Pexp_poly ({pexp_desc= Pexp_constraint (e, Some t, []); _}, None) ->
+          m.expr m {exp with pexp_desc= Pexp_poly (e, Some t)}
+      | Pexp_constraint (exp1, None, _ :: _) when erase_jane_syntax ->
+          (* When erasing jane syntax, if [Pexp_constraint] was only
+             constraining based on modes, remove the node entirely instead of
+             just making the modes list empty *)
+          m.expr m exp1
+      | Pexp_constraint (e, Some {ptyp_desc= Ptyp_poly ([], _t); _}, []) ->
+          m.expr m e
+      | Pexp_sequence
+          ( exp1
+          , { pexp_desc= Pexp_sequence (exp2, exp3)
+            ; pexp_loc= loc2
+            ; pexp_attributes= attrs2
+            ; _ } ) ->
+          m.expr m
+            (Exp.sequence ~loc:loc1 ~attrs:attrs1
+               (Exp.sequence ~loc:loc2 ~attrs:attrs2 exp1 exp2)
+               exp3 )
+      | Pexp_function
+          ( ps
+          , c
+          , Pfunction_body {pexp_desc= Pexp_function (ps', None, b'); _} ) ->
+          m.expr m {exp with pexp_desc= Pexp_function (ps @ ps', c, b')}
+      | Pexp_newtype (l, {pexp_desc= Pexp_constraint (exp1, Some ty, []); _})
+        ->
+          (* It's strange that this creates an illegal [Pexp_function] node
+             (with a single [Pparam_newtype] parameter and a
+             [Pfunction_body]), but it's necessary to normalize against a
+             strange ast change in [source.ml]. This can probably be deleted
+             when we merge with 5.2 ocamlformat *)
+          m.expr m
+            { exp with
+              pexp_desc=
+                Pexp_function
+                  ( [ { pparam_loc= l.loc
+                      ; pparam_desc= Pparam_newtype (l, None) } ]
+                  , Some
+                      {mode_annotations= []; type_constraint= Pconstraint ty}
+                  , Pfunction_body exp1 ) }
+      | Pexp_function
+          ( ps
+          , c
+          , Pfunction_body {pexp_desc= Pexp_constraint (exp1, Some ty, []); _}
+          ) ->
+          let c =
+            match c with
+            | Some c -> Some {c with type_constraint= Pconstraint ty}
```

Comment:
I'm pretty confused about what is going on here - we're throwing away the constraint on the outside in favor of the contraint on the inside.  Why is that a sensible/necessary thing to do?  I think we could miss various ocamlformat bugs this way.

Reply:
ocamlformat currently moves constraints around on functions, causing the ast to change. So `Normalize_std_ast.ml` will complain that `ast changed` if we don't do this transformation here.

Reply:
I think it may be possible to add `when Option.is_none c` here, which would fix your concern